client/state/action-types.js

```diff
@@ -78,6 +78,8 @@ export const COMMENTS_REPLY_WRITE = 'COMMENTS_REPLY_WRITE';
 export const COMMENTS_REQUEST = 'COMMENTS_REQUEST';
 export const COMMENTS_REQUEST_FAILURE = 'COMMENTS_REQUEST_FAILURE';
 export const COMMENTS_REQUEST_SUCCESS = 'COMMENTS_REQUEST_SUCCESS';
+export const COMMENTS_TREE_RECEIVE = 'COMMENTS_TREE_RECEIVE';
```

Comment:
And do we really need to be passing in _a tree_? I don't think we even are. It's more like we're passing in comment relationships or comment ids. I would feel okay if we made the reducer interface one that looks for something like `{ commentId, parentId }` where `parentId` could also be `null`/`undefined`. _it_ then builds the tree.

My thinking goes like this: when we hook up the web socket these will be coming in one at a time but we still want to make sure that the tree builds properly and we don't want to duplicate a bunch of code or skew logic.

```js
const commentTree = ( state = {}, action ) => {
	if ( COMMENT_ADD === type ) {
		return insertComment( state, action.comment.id, action.comment.parentId );
	}

	if ( COMMENT_RELATIONSHIPS_ADD === action.type ) {
		return action.comments.reduce( ( state, [ id, parentId ] ) => insertComment( state, id, parentId ), state );
	}

	if ( COMMENT_DELETE === type ) {
		return removeComment( state, action.commentId );
	}

	return state;
}
```

and in that example `insertComment( prevTree, commentId, parentId )` and `removeComment( prevTree, commentId )` are the functions which traverse the tree and insert or remove the appropriate nodes. maybe we don't even worry about this in state, leaving just a plain list and leave the tree for render-time, but I think actually the tree would be more efficient in state anyway.

That was a lot so I hope it wasn't too much given the change. I'm just not entirely sure that inserting a tree is our main idea, though maybe actually it is… 🤔 each id pair could be seen as smaller trees in and of themselves, where the data layer builds the trees it can and then inserts them…